neps/nep-0605-sharded-contracts.md

```diff
@@ -0,0 +1,182 @@
+- Storage namespace: The account balance state that is stored locally on the users' account needs to be protected.  Users cannot be allowed to modify this state directly because that will allow a malicious user the ability to corrupt the application state.  In the FT contract example, a malicious user could simply increase their balance and maliciously mint tokens.
+- Any account can send messages to any other account on NEAR.  In order to ensure safe FT transfers, there are some access control issues we will need to address.
+- Finally, there are some concerns about how a sharded contract will upgrade itself and the related state.
+
+### Storage namespaces
+
+The precise problem that we need to solve is the following:
+
+- When a sharded contract code is deployed on an account, it wants to be able to create some state on that account.
+- Only the sharded contract code should be able to make modifications to this state.
+- It should not be possible for the account to make modifications to this state.  Even after the account removes the sharded contract code.
+- Open question: can the account remove the state without being allowed to make modifications to the state?  In theory, we can allow this but then the owner of the sharded contract has to be aware of the impact of such changes.
+
+Turns out that the NEAR protocol is already solving a version of this problem.  State from all accounts is stored together in the same `Trie` data structure.  So in theory, if one account could figure out under what key another account has stored some value, it could issue a `storage_write` for that key and make unauthorised modifications to it.  To prevent this type of attack, when accounts issue a `storage_write`, the protocol prepends the `AccountId` of the account that is currently executing to the key.  This creates a unique namespace for each account that is inaccessible to all other accounts.
+
+We propose a similar approach for sharded contracts as well.  Instead of using the `AccountId` namespace, when an account starts using a sharded contract code, the protocol can create a new `ShardedContractInstanceId` for it.  This id can then be prepended to all storage accesses and just like above, this creates an isolated namespace for that specific sharded contract that the account cannot access.
+
+If an account stops using a sharded contract code (e.g. by deploying another contract code on itself), then the associated `ShardedContractInstanceId` can be removed.
```

Comment:
I am not sure what this means. The sharded contract code should be for a specific namespace and there can be many namespaces. They are independent from what contract is deployed on the account itself